cmd/swarm-rafttool/dump.go

```diff
@@ -0,0 +1,223 @@
+			fmt.Printf("Entry Index=%d, Term=%d, Type=%s:\n", ent.Index, ent.Term, ent.Type.String())
+			switch ent.Type {
+			case raftpb.EntryConfChange:
+				cc := &raftpb.ConfChange{}
+				err := proto.Unmarshal(ent.Data, cc)
+				if err != nil {
+					return err
+				}
+
+				fmt.Println("Conf change type:", cc.Type.String())
+				fmt.Printf("Node ID: %x\n\n", cc.NodeID)
+
+			case raftpb.EntryNormal:
+				r := &api.InternalRaftRequest{}
+				err := proto.Unmarshal(ent.Data, r)
+				if err != nil {
+					return err
+				}
+
+				if err := proto.MarshalText(os.Stdout, r); err != nil {
+					return err
+				}
+				fmt.Println()
+			}
+		}
+	}
+
+	return nil
+}
+
+func dumpSnapshot(swarmdir, unlockKey string) error {
+	_, snapshot, err := loadData(swarmdir, unlockKey)
+	if err != nil {
+		return err
+	}
+
+	if snapshot == nil {
+		return errors.New("no snapshot found")
+	}
+
+	s := &api.Snapshot{}
+	if err := proto.Unmarshal(snapshot.Data, s); err != nil {
+		return err
+	}
+	if s.Version != api.Snapshot_V0 {
+		return fmt.Errorf("unrecognized snapshot version %d", s.Version)
+	}
+
+	fmt.Println("Active members:")
+	for _, member := range s.Membership.Members {
+		fmt.Printf(" NodeID=%s, RaftID=%x, Addr=%s\n", member.NodeID, member.RaftID, member.Addr)
+	}
+	fmt.Println()
+
+	fmt.Println("Removed members:")
+	for _, member := range s.Membership.Removed {
+		fmt.Printf(" RaftID=%x\n", member)
+	}
+	fmt.Println()
+
+	fmt.Println("Objects:")
+	if err := proto.MarshalText(os.Stdout, &s.Store); err != nil {
+		return err
+	}
+	fmt.Println()
+
+	return nil
+}
+
+func dumpObject(swarmdir, unlockKey, objType, objID string) error {
+	memStore := store.NewMemoryStore(nil)
+	defer memStore.Close()
+
+	walData, snapshot, err := loadData(swarmdir, unlockKey)
+	if err != nil {
+		return err
+	}
+
+	if snapshot != nil {
+		var s api.Snapshot
+		if err := s.Unmarshal(snapshot.Data); err != nil {
+			return err
+		}
+		if s.Version != api.Snapshot_V0 {
+			return fmt.Errorf("unrecognized snapshot version %d", s.Version)
+		}
+
+		if err := memStore.Restore(&s.Store); err != nil {
+			return err
+		}
+	}
+
+	for _, ent := range walData.Entries {
+		if snapshot != nil && ent.Index <= snapshot.Metadata.Index {
+			continue
+		}
+
+		if ent.Type != raftpb.EntryNormal {
+			continue
+		}
+
+		r := &api.InternalRaftRequest{}
+		err := proto.Unmarshal(ent.Data, r)
+		if err != nil {
+			return err
+		}
+
+		if r.Action != nil {
+			if err := memStore.ApplyStoreActions(r.Action); err != nil {
+				return err
+			}
+		}
+	}
+
+	var object proto.Message
+	memStore.View(func(tx store.ReadTx) {
+		switch objType {
+		case "node":
+			object = store.GetNode(tx, objID)
+		case "service":
+			object = store.GetService(tx, objID)
+		case "task":
+			object = store.GetTask(tx, objID)
+		case "network":
+			object = store.GetNetwork(tx, objID)
+		case "cluster":
+			object = store.GetCluster(tx, objID)
+		case "secret":
+			object = store.GetSecret(tx, objID)
+		}
```

Comment:
Also wondering if name would make more sense than ID, since the ID is random (and the user would have to look up the item by ID first)